### ISSUE TITLE
pip install "soup-cli[mlx]" cannot train: the MLX path imports datasets + TRL before reaching the MLX branch

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#394]
Split out of #392 (reported by @armanbot-jpg), because the fix is either a real dependency change or a documentation correction and those are different discussions.

## What

`docs/` presents `pip install "soup-cli[mlx]"` as the install path for Apple Silicon fine-tuning. It is not sufficient: `soup train` reaches `datasets`, then TRL, before control ever gets to the MLX branch, so the run dies on an import that has nothing to do with MLX. In practice a user needs the full `[train]` extra — which pulls **torch** onto a machine that was going to use MLX precisely so it would not need torch.

Reported on macOS arm64, Python 3.12, mlx 0.30.x / mlx-lm 0.31.x.

## Why it is worth fixing rather than documenting away

The whole proposition of `backend: mlx` is "you do not need the CUDA stack". An install path that pulls torch anyway keeps the dependency it exists to avoid, and the doc currently promises otherwise — so a user following it hits an ImportError that names a package they never asked for.

## Two candidate fixes, and they are not equivalent

1. **Make `[mlx]` actually sufficient.** Push the `datasets` / TRL imports behind the backend branch so an MLX run never touches them, and add whatever `[mlx]` genuinely needs. This is the fix that keeps the promise, and it is the larger one — it means auditing the import path from `soup train` to `MLXSFTTrainerWrapper.setup()` for anything torch-shaped.
2. **Correct the docs** to say `[mlx]` is additive to `[train]`, not a replacement. Honest, one line, and it concedes the proposition.

(1) is the right answer if the import audit is small; (2) is the right answer if it is not. Whoever picks this up should measure before choosing — the deciding fact is how many of the imports between the CLI entry point and the MLX branch are structural rather than incidental.

## Acceptance criteria

- [ ] a clean venv with **only** `pip install "soup-cli[mlx]"` either completes a tiny MLX SFT run, or the docs no longer claim it can
- [ ] if (1): a test that pins the MLX path does not import torch/TRL — the same shape as `tests/test_cli_startup_is_light.py`, which asserts the *runtime* property (`"torch" not in sys.modules`) rather than a syntactic one
- [ ] `docs/backends-and-ops.md` and any other page naming the `[mlx]` extra agree with whichever answer is taken

**Comment by MakazhanAlpamys** [comment]
## Measured, and it does not reproduce on `main` — asking for the traceback before changing anything

Before choosing between fix (1) and fix (2) I ran the measurement the acceptance criteria call for. The result argues for neither, because the failure did not occur.

**Method.** Block the entire training stack by import name — `torch`, `datasets`, `trl`, `peft`, `accelerate`, `bitsandbytes` — then drive the real CLI (`CliRunner`, not a direct call into the wrapper) with an MLX config and a local JSONL dataset:

```
exit code: 1
exception: ImportError("MLX backend requires the 'mlx' and 'mlx-lm' packages.
                       Install with: pip install \"soup-cli[mlx]\"")
blocked import attempts: ['torch', 'torch']
last line printed:      Setting up model + trainer...
```

Reading that in order:

- **`datasets` and `trl` are never imported at all** — zero attempts, not "attempted and handled". The config parsed, the dataset loaded from local JSONL, `resolve_trainer` returned `MLXSFTTrainerWrapper`, the setup panel rendered, and control reached `setup()`.
- **The two `torch` attempts are both handled** — they come from the device probe, which is why the run prints *"4bit quantization is not supported on CPU. Switching to quantization: none"* instead of dying.
- **The only thing that stopped it was `_require_mlx()`**, raising the correct error naming the correct install command — i.e. exactly the intended failure on a box with no `mlx`, which this one is.

Separately, the `[mlx]` extra does exist and is `["mlx>=0.20.0", "mlx-lm>=0.31.3"]`, so `pip install "soup-cli[mlx]"` is not a no-op that silently installs the base package.

## So the issue as filed is not confirmed, and I would rather ask than guess

@armanbot-jpg — your other two findings were exact, and one of them (#392) was a real silent-corruption bug I would not have found. So I am not dismissing this; I am reporting that my instrument disagrees with it, which usually means the instrument is measuring the wrong thing.

Three ways the difference could be real, and any one of them changes the fix:

1. **A different command.** This test drives `soup train`. If you hit it on `soup data …`, `soup profile`, or something else in the same session, that path may well import `datasets` while `train` does not.
2. **A different import.** Blocking by top-level name catches `import datasets` and `from datasets import …`, but not a package that vendors it under another name. The traceback would say immediately.
3. **The tree.** You were on `9f57b7c` (14 Aug 15:25); `main` is 7 commits ahead now. Nothing in those looks import-related, but it is cheap to rule out.

**What would settle it in one paste:** the full traceback, and the exact command line. `pip list | grep -iE "torch|datasets|trl|mlx"` from that venv would help too — if `torch` is present in an `[mlx]`-only environment, something is pulling it and that is the actual bug.

Leaving this open and unassigned meanwhile. It is not worth restructuring imports that are already lazy on the strength of a failure neither of us can currently reproduce, and it is not worth downgrading the docs to match a limitation that may not exist.

**Comment by riviangeralt** [comment]
Data point on the hang variant (not just the ImportError) — repro:

- macOS arm64 (M1), Python 3.12.13 (uv-managed venv), `soup-cli[mlx]` 0.73.2, `mlx` 0.32.0, `mlx-lm` 0.31.3
- `pip install "soup-cli[mlx]"` alone → `soup train` reported `Device: CPU (no GPU detected)` and silently downgraded `quantization: 4bit` → `none` (`detect_device()` in `gpu.py` only checks `torch.cuda`/`torch.backends.mps`, so with no torch installed it always falls through to the CPU branch — cosmetic wrt this issue, but a second symptom of the same "MLX-only install isn't self-sufficient" root cause worth folding into the fix).
- Installed `torch` (CPU/MPS build) as a workaround → device panel then correctly reported `Apple Silicon (MPS)`, `Quant: 4bit`.
- With torch present, `soup train --config soup.yaml --yes` no longer ImportErrors, but **hangs** instead — reproduced 3/3 runs, stalling at slightly different points each time (once at `Calculating loss...: 0/1` during eval, twice somewhere between the `target_modules: auto cannot be resolved...` warning and the `Starting training..., iters: N` line). CPU usage during the hangs was low/bursty (not a tight compute loop), consistent with the run being stuck initializing something in the `datasets`/TRL/transformers import path this issue describes, rather than in the actual MLX training step.
- Base model: local dir with pre-converted MLX 4-bit weights (`mlx-community`-style, `base_model: Qwen/Qwen2.5-Coder-7B-Instruct`), `backend: mlx`, `task: sft`, tiny dataset (4 alpaca-format examples).
- Killed and retried 3x with minor config changes (disabling `val_split`) — same class of stall each time, never completed.

Consistent with candidate fix (1) in the issue — the `datasets`/TRL import path being reached at all for an MLX run seems to be causing more than just the missing-torch crash.

**Comment by MakazhanAlpamys** [comment]
@riviangeralt — thank you, this is the second independent report and it moves the issue substantially. Two things in it, and the second is a defect nobody had named yet.

## 1. The hang

Reproduced 3/3 with torch present, stalling at slightly different points and with low/bursty CPU rather than a tight loop. That is a materially different symptom from the ImportError this issue was filed for, and it is one my own measurement could not have found: I tested by **blocking** `torch` / `datasets` / `trl` by import name and driving the real CLI, which reaches `_require_mlx()` cleanly and never touches them. A run where those packages are **present** takes a different path entirely, and blocking them made that path unreachable.

So my earlier "does not reproduce" stands only for the narrow claim it tested — that the MLX path *imports* `datasets`/TRL. It clearly does not settle whether that path is what hangs when they are importable.

What would pin it: `py-spy dump --pid <pid>` (or `faulthandler.dump_traceback_later(30)`) on a stalled run. A stack from the hang tells us in one paste whether it is stuck in an import, in `datasets` Arrow/multiprocessing setup, or in MLX itself — and those three want completely different fixes.

## 2. `detect_device()` does not know MLX — and it silently downgrades quantization

This is the part I want to split out, because it is concrete, checkable, and independent of the hang:

> `soup train` reported `Device: CPU (no GPU detected)` and silently downgraded `quantization: 4bit` → `none`

`detect_device()` checks `torch.cuda` and `torch.backends.mps` only. On an Apple Silicon box running `backend: mlx`, the GPU MLX will actually use is invisible to it — so the pre-flight prints "no GPU detected" and the 4-bit → none rewrite fires on a machine that was going to run 4-bit on the GPU.

That is a **silent config change on a correct configuration**, which is a failure class this project treats as worse than a crash: the run completes, the panel says something false, and the model is not the one that was asked for. It is also exactly the shape of the MLX bug fixed last week (#392) — the loss curve looks fine and the artifact is not what you configured.

Filing that separately today with your reproduction credited. It is fixable without settling the hang.

---

Both of your reports have been precise about the difference between what you observed and what you inferred, which is why they are actionable. If you can get a stack dump from a stalled run, that closes the remaining half.

**Comment by MakazhanAlpamys** [comment]
Status update — **this issue stays open**, deliberately.

[#431](https://github.com/MakazhanAlpamys/Soup/pull/431) by **[@Shutaru](https://github.com/Shutaru)** merged as [`e399868`](https://github.com/MakazhanAlpamys/Soup/commit/e399868), but it is scoped as *hardening*, not as a fix for either variant reported here. The author retitled it to say so once that came out in review, and there is no auto-closing keyword in the PR or its commits.

## What was measured

With nothing blocked, a real `backend: mlx` CLI run imports **only `torch` and `numpy`**. `datasets`, `trl`, `peft`, `accelerate`, `bitsandbytes` and `transformers` are all absent; `torch`'s single traced call site is `utils/gpu.py:100`, inside `try/except ImportError`, so it degrades cleanly. This matches my own earlier measurement in this thread.

So **the ImportError variant does not reproduce as filed.**

## What is still open — and it is the substantive half

@riviangeralt's **hang** on macOS arm64, reproduced 3/3 **with torch present**. Nothing in #431 changes behaviour in that case, and there is still no stack dump. That is what this issue now tracks.

If anyone can get a `faulthandler.dump_traceback_later()` or a `py-spy dump` off a hung run, that would move it further than anything else. Without it we are guessing at where it blocks.

## What #431 did deliver

`sft.py` is import-light *today*, and the PR stops that from being load-bearing: the Transformers SFT wrapper import now sits behind the non-MLX dispatch branch, with a regression test that blocks those roots and fails on a straight revert. It also adds an Apple Silicon `mlx-smoke` CI job that asserts mlx is genuinely importable rather than skipping green — the previous shape would have reported success having executed nothing.

Separately, `docs/backends-and-ops.md` now names #423 on the standalone `[mlx]` install, since that path silently rewrites `quantization: 4bit` to `none` until #423 lands.

**Comment by MakazhanAlpamys** [comment]
Correction to my note above, and it is my mistake rather than anyone else's.

**This issue was auto-closed by my own merge commit**, seconds after I wrote that it would stay open. Reopened now.

The cause is worth recording, because it is a trap anyone here can hit. My squash body contained the sentence:

> Deliberately does NOT close #394.

GitHub's issue-linking parser matches the substring `close #394` and **does not read the negation**. So the sentence I wrote specifically to prevent the issue from closing is what closed it.

I had checked the right thing and drawn the wrong conclusion: I verified there was no closing keyword in the **PR body or the PR's commits** — which was true, and is why I told @Shutaru the merge could not auto-close it. What I did not check was the **squash commit message I wrote myself at merge time**. That message is a third place a keyword can live, and it is the one I authored last and reviewed least.

Nothing about the technical status changes: the ImportError variant does not reproduce as filed, the macOS arm64 hang (3/3, torch present) is untouched and still has no stack dump, and that hang is what this issue tracks.

For the record, so nobody repeats it: to reference an issue from a commit without closing it, write the bare number (`#394`) or `re #394` / `refs #394`. Never write a closing verb next to it, **even negated** — `does not close #394`, `should not fix #394` and `this doesn't resolve #394` all close the issue.

**Comment by MakazhanAlpamys** [comment]
Closing — fix (1) was taken, not the documentation retreat, and all three acceptance boxes are covered.

**1. "A clean venv with only `pip install "soup-cli[mlx]"` completes a tiny MLX SFT run."** This is now a CI job rather than a claim. `.github/workflows/ci.yml` runs `mlx-smoke` on `macos-14`, and it does the three things this issue asked for, in order:

```yaml
- run: pip install -e ".[mlx]" pytest
- run: |
    python -c "import mlx.core, mlx_lm; ..."
    python -c "... names=('torch','trl','datasets','peft','accelerate','bitsandbytes'); present=[...]; assert not present, ..."
- run: pytest -o addopts= -m smoke -k mlx_sft_smoke -q tests/test_smoke_train.py
```

It installs *only* the extra, asserts the six training-stack packages are absent from the environment at all, and then trains. The job comment records why it is kept out of the main matrix: `.[dev]` deliberately installs the PyTorch/TRL stack and would hide an accidental MLX dependency — which is the failure mode that would have made this test worthless.

**2. "A test that pins the MLX path does not import torch/TRL — the runtime property, not a syntactic one."** `tests/test_issue394_mlx_runtime_isolation.py::test_mlx_cli_route_does_not_touch_transformers_training_stack`. It installs a `MetaPathFinder` that *raises* on the blocked modules rather than grepping for import statements, so an import reached through any indirection still fails. That is stricter than the `"torch" not in sys.modules` shape the issue asked for by name: absence-after-the-fact can be satisfied by an import that happened and was cleaned up, while a blocking finder cannot.

**3. "`docs/backends-and-ops.md` and any other page naming the extra agree."** `docs/backends-and-ops.md:85` now says `[mlx]` is a standalone install, scoped to SFT with local JSONL / JSON / CSV data. That qualifier is the honest part — it states where the promise holds instead of restating it unconditionally.

Green in the full suite locally (18087 passed), and `mlx-smoke` has been passing on recent PRs.
